File: byul_demo/wrapper/modules/route.py
from ffi_core import ffi, C
import weakref

from coord import c_coord
from coord_list import c_coord_list
from coord_hash import c_coord_hash

from enum import IntEnum

class RouteDir(IntEnum):
    UNKNOWN = 0
    RIGHT = 1
    TOP_RIGHT = 2
    TOP = 3
    TOP_LEFT = 4
    LEFT = 5
    DOWN_LEFT = 6
    DOWN = 7
    DOWN_RIGHT = 8
    COUNT = 9

ffi.cdef("""
typedef enum e_route_dir {
    ROUTE_DIR_UNKNOWN, 
    ROUTE_DIR_RIGHT,
    ROUTE_DIR_TOP_RIGHT,
    ROUTE_DIR_TOP,
    ROUTE_DIR_TOP_LEFT,
    ROUTE_DIR_LEFT,
    ROUTE_DIR_DOWN_LEFT,
    ROUTE_DIR_DOWN,
    ROUTE_DIR_DOWN_RIGHT,
    ROUTE_DIR_COUNT
} route_dir_t;

struct s_route {
    coord_list_t* coords;
    coord_list_t* visited_order;
    coord_hash_t* visited_count;
    float cost;
    bool success;
    int total_retry_count;

    float avg_vec_x;
    float avg_vec_y;
    int vec_count;
};

typedef struct s_route route_t;

/** 생성 및 해제 **/
route_t* route_new(void);
route_t* route_new_full(float cost);
void  route_free(route_t* p);

/** 복사 및 비교 **/
route_t* route_copy(const route_t* p);
uintptr_t route_hash(const route_t* a);
int route_equal(const route_t* a, const route_t* b);

/** 기본 정보 **/
void  route_set_cost(route_t* p, float cost);
float route_get_cost(const route_t* p);
void  route_set_success(route_t* p, int success);
int   route_get_success(const route_t* p);

/** 좌표 리스트 접근 **/
const coord_list_t* route_get_coords(const route_t* p);

/** 방문 로그 **/
const coord_list_t* route_get_visited_order(const route_t* p);
const coord_hash_t*  route_get_visited_count(const route_t* p);

int route_get_total_retry_count(const route_t* p);

void route_set_total_retry_count(route_t* p, int retry_count);

/** 좌표 조작 **/
int  route_add_coord(route_t* p, const coord_t* c);
void route_clear_coords(route_t* p);
const coord_t* route_get_last(const route_t* p);
const coord_t* route_get_coord_at(const route_t* p, int index);
int   route_length(const route_t* p);

/** 방문 조작 **/
int  route_add_visited(route_t* p, const coord_t* c);
void route_clear_visited(route_t* p);

/** 병합 및 편집 **/
void route_append(route_t dest, const route_t* src);
void route_append_nodup(route_t dest, const route_t* src);

void route_insert(route_t* p, int index, const coord_t* c);
void route_remove_at(route_t* p, int index);
void route_remove_value(route_t* p, const coord_t* c);
int  route_contains(const route_t* p, const coord_t* c);
int  route_find(const route_t* p, const coord_t* c);
void route_slice(route_t* p, int start, int end);

/** 출력 및 디버깅 **/
void route_print(const route_t* p);

/** 방향 계산 **/
coord_t* route_make_direction(route_t* p, int index);
route_dir_t route_get_direction_by_coord(const coord_t* dxdy);
route_dir_t route_get_direction_by_index(route_t* p, int index);
route_dir_t route_calc_average_facing(route_t* p, int history);
float route_calc_average_dir(route_t* p, int history);

coord_t* direction_to_coord(route_dir_t route_dir);

/** 방향 변화 판단 **/
int route_has_changed(
    route_t* p, const coord_t* from,
    const coord_t* to, float angle_threshold_deg);

int route_has_changed_with_angle(
    route_t* p, const coord_t* from,
    const coord_t* to, float angle_threshold_deg,
    float* out_angle_deg);

int route_has_changed_by_index(
    route_t* p, int index_from,
    int index_to, float angle_threshold_deg);

int route_has_changed_with_angle_by_index(
    route_t* p, int index_from, int index_to,
    float angle_threshold_deg, float* out_angle_deg);

/** 평균 벡터 누적 **/
void route_update_average_vector(
    route_t* p, const coord_t* from, const coord_t* to);

void route_update_average_vector_by_index(
    route_t* p, int index_from, int index_to);

route_dir_t calc_direction(
    const coord_t* start, const coord_t* goal);

/// @brief goal → start 방향으로 came_from을 따라 경로를 복원하여 route에 채운다.
/// @param route 결과 경로 구조체 (output)
/// @param came_from coord_hash_t* (coord* → coord*)
/// @param start 시작 좌표
/// @param goal 도착 좌표
/// @return 성공 여부 (true: 경로 복원 성공, false: 실패)
bool route_reconstruct_path(
    route_t* route, const coord_hash_t* came_from,
    const coord_t* start, const coord_t* goal);

""")

class c_route:
    def __init__(self, raw_ptr=None, cost=None, own=False):
        if raw_ptr:
            self._c = raw_ptr
            self._own = own
        elif cost is not None:
            self._c = C.route_new_full(cost)
            self._own = True
        else:
            self._c = C.route_new()
            self._own = True

        if not self._c:
            raise MemoryError("route allocation failed")

        self._finalizer = weakref.finalize(self, C.route_free, self._c)

    # ───── 기본 정보 ─────
    def cost(self):
        return C.route_get_cost(self._c)

    def set_cost(self, cost: float):
        C.route_set_cost(self._c, cost)

    def is_success(self):
        return bool(C.route_get_success(self._c))

    def set_success(self, success: bool):
        C.route_set_success(self._c, int(success))

    def retry_count(self):
        return C.route_get_total_retry_count(self._c)

    def set_retry_count(self, count: int):
        C.route_set_total_retry_count(self._c, count)

    # ───── 경로 좌표 ─────
    def coords(self):
        return c_coord_list(raw_ptr=C.route_get_coords(self._c), own=False)

    def add_coord(self, coord: c_coord):
        return C.route_add_coord(self._c, coord.ptr())

    def clear_coords(self):
        C.route_clear_coords(self._c)

    def last(self):
        ptr = C.route_get_last(self._c)
        return c_coord(raw_ptr=ptr) if ptr != ffi.NULL else None

    def coord_at(self, index):
        ptr = C.route_get_coord_at(self._c, index)
        return c_coord(raw_ptr=ptr) if ptr != ffi.NULL else None

    def length(self):
        return C.route_length(self._c)

    # ───── 방문 로그 ─────
    def visited_order(self):
        return c_coord_list(raw_ptr=C.route_get_visited_order(self._c), own=False)

    def visited_count(self):
        return c_coord_hash(raw_ptr=C.route_get_visited_count(self._c), own=False)

    def add_visited(self, coord: c_coord):
        return C.route_add_visited(self._c, coord.ptr())

    def clear_visited(self):
        C.route_clear_visited(self._c)

    # ───── 경로 병합 및 편집 ─────
    def append(self, other: 'c_route', nodup=False):
        if nodup:
            C.route_append_nodup(self._c, other._c)
        else:
            C.route_append(self._c, other._c)

    def insert(self, index, coord: c_coord):
        C.route_insert(self._c, index, coord.ptr())

    def remove_at(self, index):
        C.route_remove_at(self._c, index)

    def remove_value(self, coord: c_coord):
        C.route_remove_value(self._c, coord.ptr())

    def contains(self, coord: c_coord):
        return bool(C.route_contains(self._c, coord.ptr()))

    def find(self, coord: c_coord):
        return C.route_find(self._c, coord.ptr())

    def slice(self, start, end):
        C.route_slice(self._c, start, end)

    # ───── 방향 처리 ─────
    def look_at(self, index):
        ptr = C.route_make_direction(self._c, index)
        return c_coord(raw_ptr=ptr) if ptr != ffi.NULL else None

    def get_direction_by_coord(self, dxdy: c_coord):
        return RouteDir(C.route_get_direction_by_coord(dxdy.ptr()))

    def get_direction_by_index(self, index):
        return RouteDir(C.route_get_direction_by_index(self._c, index))

    def calc_average_facing(self, history):
        return RouteDir(C.route_calc_average_facing(self._c, history))

    def calc_average_dir(self, history):
        return C.route_calc_average_dir(self._c, history)

    # ───── 방향 변화 판단 ─────
    def has_changed(self, from_coord, to_coord, angle_threshold):
        return bool(C.route_has_changed(self._c, from_coord.ptr(), to_coord.ptr(), angle_threshold))

    def has_changed_with_angle(self, from_coord, to_coord, angle_threshold):
        out = ffi.new("float*")
        changed = C.route_has_changed_with_angle(self._c, from_coord.ptr(), to_coord.ptr(), angle_threshold, out)
        return bool(changed), out[0]

    def has_changed_by_index(self, index_from, index_to, angle_threshold):
        return bool(C.route_has_changed_by_index(self._c, index_from, index_to, angle_threshold))

    def has_changed_with_angle_by_index(self, index_from, index_to, angle_threshold):
        out = ffi.new("float*")
        changed = C.route_has_changed_with_angle_by_index(self._c, index_from, index_to, angle_threshold, out)
        return bool(changed), out[0]

    def update_average_vector(self, from_coord, to_coord):
        C.route_update_average_vector(self._c, from_coord.ptr(), to_coord.ptr())

    def update_average_vector_by_index(self, index_from, index_to):
        C.route_update_average_vector_by_index(self._c, index_from, index_to)

    def reconstruct_path(self, came_from: c_coord_hash, start: c_coord, goal: c_coord):
        return bool(C.route_reconstruct_path(self._c, came_from.ptr(), start.ptr(), goal.ptr()))

    def print(self):
        C.route_print(self._c)

    def ptr(self):
        return self._c

    def __repr__(self):
        return f"c_route(len={self.length()}, cost={self.cost():.2f}, success={self.is_success()})"

    def __del__(self):
        if self._own and self._finalizer and self._finalizer.alive:
            self._finalizer()

    def close(self):
        if self._own and self._finalizer and self._finalizer.alive:
            self._finalizer()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

def direction_to_coord(direction: RouteDir) -> c_coord:
    ptr = C.direction_to_coord(direction)
    return c_coord(raw_ptr=ptr) if ptr != ffi.NULL else None

def calc_direction(start: c_coord, goal: c_coord) -> RouteDir:
    return RouteDir(C.calc_direction(start.ptr(), goal.ptr()))

File: byul_demo/routefinder/route/route.c
#include "internal/route.h"
#include "internal/core.h"
#include <math.h>

#define EPSILON 1e-6f

static const int ROUTE_DIRECTION_VECTORS[9][2] = {
    {  0,  0 },  // UNKNOWN
    {  1,  0 },  // RIGHT
    {  1, -1 },  // TOP_RIGHT
    {  0, -1 },  // TOP
    { -1, -1 },  // TOP_LEFT
    { -1,  0 },  // LEFT
    { -1,  1 },  // DOWN_LEFT
    {  0,  1 },  // DOWN
    {  1,  1 },  // DOWN_RIGHT
};


route route_new(void) {
    return route_new_full(0.0f);
}

route route_new_full(gfloat cost) {
    route p = g_malloc0(sizeof(route_t));
    if (!p) return NULL;

    p->coords = NULL;
    p->cost = cost;
    p->success = FALSE;

    p->visited_order = NULL;

    p->visited_count = g_hash_table_new_full(
        (GHashFunc) coord_hash, 
        (GEqualFunc) coord_equal,
        (GDestroyNotify) coord_free, 
        NULL
    );

    return p;
}

void route_free(const route p) {
    if (!p) return;
    if (p->coords) g_list_free_full(p->coords, (GDestroyNotify) coord_free);
    if (p->visited_count) g_hash_table_destroy(p->visited_count);
    if (p->visited_order) {
        g_list_free_full(p->visited_order, (GDestroyNotify) coord_free);
    }

    g_free(p);
}

guint route_hash(const route p) {
    return (guint)(guintptr)p;
}

gboolean route_equal(const route a, const route b) {
    if (!a || !b) return FALSE;
    return g_direct_equal(a, b);    
}

route route_copy(const route p) {
    if (!p) return NULL;
    route c = route_new();
    // c->coords = p->coords;
    c->coords = g_list_copy_deep(p->coords,
        (GCopyFunc) coord_copy, NULL);
    c->cost = p->cost;
    c->success = p->success;
    // if (p->visited) c->visited = p->visited;
    if (p->visited_count) {
        // 이미 new 시점에 c->visited는 메모리 할당되어있다
        g_hash_table_destroy(c->visited_count);
        c->visited_count = NULL;
        c->visited_count = g_hash_table_copy_deep(p->visited_count, 
            (GHashFunc) coord_hash,
            (GEqualFunc) coord_equal,
            (GCopyFunc) coord_copy, 
            (GDestroyNotify) coord_free,
            NULL, NULL);
        }
    
    // if (p->visited_order) c->visited_order = p->visited_order;
    if (p->visited_order) {
        c->visited_order = g_list_copy_deep(p->visited_order,
        (GCopyFunc) coord_copy, NULL);
    }

    return c;
}

void route_set_cost(route p, gfloat cost) {
    if (p) p->cost = cost;
}

gfloat route_get_cost(const route p) {
    return p ? p->cost : 0.0f;
}

void route_set_success(route p, gboolean success) {
    if (p) p->success = success;
}

gboolean route_get_success(const route p) {
    return p ? p->success : FALSE;
}

GList* route_get_coords(const route p) {     
    return p ? p->coords : NULL;
}

void route_set_coords(const route p, const GList* coords) {
    if (!p || !coords) return;
    if (p->coords) {
        g_list_free_full(p->coords, (GDestroyNotify) coord_free);
    }
    p->coords = g_list_copy_deep((GList*)coords,
        (GCopyFunc) coord_copy, NULL);
}

GHashTable* route_get_visited_count(const route p) {
    return p ? p->visited_count : NULL;
}

void route_set_visited_count(route p, const GHashTable* visited_count) {
    if (!p) return;
    if (p->visited_count) g_hash_table_destroy(p->visited_count);
    // p->visited = visited;
    p->visited_count = g_hash_table_copy_deep((GHashTable*)visited_count,
        (GHashFunc) coord_hash,
        (GEqualFunc) coord_equal,
        (GCopyFunc) coord_copy,
        (GDestroyNotify) coord_free,
        NULL, NULL);
}

GList* route_get_visited_order(const route p) {
    return p ? p->visited_order : NULL;    
}

void route_set_visited_order(route p, const GList* visited_order) {
    if (!p || !visited_order) return;
    if (p->visited_order) {
        g_list_free_full(p->visited_order, (GDestroyNotify) coord_free);
    }
    p->visited_order = g_list_copy_deep((GList*)visited_order, 
        (GCopyFunc) coord_copy, NULL);
}

gboolean route_add_coord(route p, const coord c) {
    if (!p || !c) return FALSE;

    coord copy = coord_copy(c);
    if (!copy) return FALSE;

    p->coords = g_list_append(p->coords, copy);
    return TRUE;
}

void route_clear_coords(route p) {
    if (!p || !p->coords) return;
    g_list_free_full(p->coords, (GDestroyNotify)coord_free);
    p->coords = NULL;
}

gboolean route_add_visited(route p, const coord c) {
    if (!p || !c) return FALSE;

    p->visited_order = g_list_append(p->visited_order, coord_copy(c));

    gpointer val = g_hash_table_lookup(p->visited_count, c);
    gint count = val ? GPOINTER_TO_INT(val) : 0;
    count++;

    gboolean result = g_hash_table_replace(
        p->visited_count,
        coord_copy(c),
        GINT_TO_POINTER(count)
    );

    return result;
}

void route_clear_visited(route p) {
    if (!p || !p->visited_count) return;
    if (p->visited_order)
        g_list_free_full(p->visited_order, (GDestroyNotify)coord_free);
    if (p->visited_count) g_hash_table_destroy(p->visited_count);
}

// void route_add_from_came_from(route p, GHashTable* came_from, 
//     const coord goal) {

//     GList* rev_route = reconstruct_route(came_from, goal);
//     for (GList* l = rev_route; l != NULL; l = l->next) {
//         route_add_coord(p, (coord)l->data);
//     }
//     g_list_free(rev_route);
// }

gint route_length(const route p) {
    if (!p || !p->coords) return 0;
    return g_list_length(p->coords);
}

coord route_get_last(const route p) {
    if (!p || !p->coords) return NULL;
    return (coord)g_list_last(p->coords)->data;
}

void route_append(route dest, const route src) {
    if (!dest || !src || !src->coords) return;

    for (GList* l = src->coords; l; l = l->next) {
        coord c = (coord)l->data;
        // route_add_coord(dest, coord_copy(c));
        route_add_coord(dest, c);
    }
}

void route_append_nodup(route dest, const route src) {
    if (!dest || !src || !src->coords) return;

    GList* tail = dest->coords;
    coord last = tail ? (coord)g_list_last(tail)->data : NULL;

    for (GList* l = src->coords; l; l = l->next) {
        coord c = (coord)l->data;

        if (last && coord_equal(last, c)) {
            continue; // 중복 방지
        }

        coord c_copy = coord_copy(c);
        dest->coords = g_list_append(dest->coords, c_copy);
        last = c_copy;
    }
}

void route_print(const route p) {
    g_print("최종 경로 : ");
    for (GList* l = route_get_coords(p); l; l = l->next) {
            coord c = (coord)l->data;
            if (l->next) {
                g_print("(%d, %d) -> ", c->x, c->y);
            } else {
                g_print("(%d, %d)", c->x, c->y);
            }
        }    
    g_print("\n");    
}

coord route_look_at(route p, int index) {
    if (!p || route_length(p) < 2) return NULL;

    int len = route_length(p);
    if (index < 0 || index >= len) return NULL;

    coord prev = NULL;
    coord curr = route_get_coord_at(p, index);
    coord next = NULL;

    if (index >= 1)
        prev = route_get_coord_at(p, index - 1);

    if (index < len - 1)
        next = route_get_coord_at(p, index + 1);

    int vx = 0;
    int vy = 0;

    if (prev) {
        vx += (curr->x - prev->x);
        vy += (curr->y - prev->y);
    }

    if (next) {
        vx += (next->x - curr->x);
        vy += (next->y - curr->y);
    }

    // 정규화 없이 단순 방향 정수 근사
    if (vx > 1) vx = 1;
    if (vx < -1) vx = -1;
    if (vy > 1) vy = 1;
    if (vy < -1) vy = -1;

    if (vx == 0 && vy == 0)
        return NULL;

    return coord_new_full(vx, vy);
}

// route_dir_t route_get_direction_by_coord(const coord dxdy) {
//     if (!dxdy) return ROUTE_DIR_UNKNOWN;

//     for (int i = 1; i <= 8; ++i) {
//         if (dxdy->x == ROUTE_DIRECTION_VECTORS[i][0] &&
//             dxdy->y == ROUTE_DIRECTION_VECTORS[i][1]) {
//             return (route_dir_t)i;
//         }
//     }

//     return ROUTE_DIR_UNKNOWN;
// }

// route_dir_t route_get_direction_by_coord(const coord dxdy) {
//     if (!dxdy || (dxdy->x == 0 && dxdy->y == 0))
//         return ROUTE_DIR_UNKNOWN;

//     gfloat best_score = -G_MAXFLOAT;
//     int best_dir = 0;

//     for (int i = 1; i <= 8; ++i) {
//         int vx = ROUTE_DIRECTION_VECTORS[i][0];
//         int vy = ROUTE_DIRECTION_VECTORS[i][1];

//         gfloat score = dxdy->x * vx + dxdy->y * vy; // 내적 (cos 유사도와 유사)

//         if (score > best_score) {
//             best_score = score;
//             best_dir = i;
//         }
//     }

//     return (route_dir_t)best_dir;
// }

route_dir_t route_get_direction_by_coord(const coord dxdy) {
    if (!dxdy || (dxdy->x == 0 && dxdy->y == 0))
        return ROUTE_DIR_UNKNOWN;

    int nx = (dxdy->x > 0) ? 1 : (dxdy->x < 0) ? -1 : 0;
    int ny = (dxdy->y > 0) ? 1 : (dxdy->y < 0) ? -1 : 0;

    for (int i = 1; i <= 8; ++i) {
        if (ROUTE_DIRECTION_VECTORS[i][0] == nx &&
            ROUTE_DIRECTION_VECTORS[i][1] == ny) {
            return (route_dir_t)i;
        }
    }

    return ROUTE_DIR_UNKNOWN;
}

route_dir_t route_get_direction_by_index(route p, int index) {
    coord dxdy = route_look_at(p, index);
    if (!dxdy) return ROUTE_DIR_UNKNOWN;
    route_dir_t result = route_get_direction_by_coord(dxdy);
    coord_free(dxdy);
    return result;
}

gboolean route_has_changed(
    route p, const coord from, const coord to, gfloat threshold_deg) {

    if (!p || !from || !to) return FALSE;

    float dx = to->x - from->x;
    float dy = to->y - from->y;
    float len = sqrtf(dx * dx + dy * dy);
    if (len < 1e-6f) return FALSE;  // 정지 상태

    float curr_vec[2] = { dx / len, dy / len };

    // 내부 벡터 히스토리 누적 or 평균 벡터 사용
    float avg_x = p->avg_vec_x;  // route 구조체에 누적 필드 있다고 가정
    float avg_y = p->avg_vec_y;

    float dot = curr_vec[0] * avg_x + curr_vec[1] * avg_y;
    if (dot > 1.0f) dot = 1.0f;
    if (dot < -1.0f) dot = -1.0f;

    float angle_rad = acosf(dot);
    float angle_deg = angle_rad * 180.0f / M_PI;

    return angle_deg > threshold_deg;
}

gboolean route_has_changed_with_angle(
    route p,
    const coord from,
    const coord to,
    gfloat angle_threshold_deg,
    gfloat* out_angle_deg
) {
    if (!p || !from || !to || !out_angle_deg)
        return FALSE;

    float dx = to->x - from->x;
    float dy = to->y - from->y;
    float length = sqrtf(dx * dx + dy * dy);
    if (length < EPSILON) {
        *out_angle_deg = 0.0f;
        return FALSE; // 정지 상태
    }

    float curr_vec_x = dx / length;
    float curr_vec_y = dy / length;

    // 평균 벡터가 아직 유효하지 않다면 (처음 입력)
    if (p->vec_count == 0) {
        p->avg_vec_x = curr_vec_x;
        p->avg_vec_y = curr_vec_y;
        p->vec_count = 1;

        *out_angle_deg = 0.0f;
        return FALSE; // 첫 입력은 비교하지 않음
    }

    // 평균 벡터 정규화
    float avg_length = sqrtf(
        p->avg_vec_x * p->avg_vec_x + p->avg_vec_y * p->avg_vec_y);

    if (avg_length < EPSILON) {
        *out_angle_deg = 0.0f;
        return FALSE;
    }

    float avg_vec_x = p->avg_vec_x / avg_length;
    float avg_vec_y = p->avg_vec_y / avg_length;

    // 내적(dot product) 계산
    float dot = curr_vec_x * avg_vec_x + curr_vec_y * avg_vec_y;
    if (dot > 1.0f) dot = 1.0f;
    if (dot < -1.0f) dot = -1.0f;

    float angle_rad = acosf(dot);
    float angle_deg = angle_rad * 180.0f / (float)M_PI;
    *out_angle_deg = angle_deg;

    gboolean changed = (angle_deg > angle_threshold_deg);

    // 평균 벡터 갱신 (단순 누적 방식)
    p->avg_vec_x += curr_vec_x;
    p->avg_vec_y += curr_vec_y;
    p->vec_count += 1;

    return changed;
}

void route_update_average_vector(
    route p,
    const coord from,
    const coord to
) {
    if (!p || !from || !to)
        return;

    float dx = to->x - from->x;
    float dy = to->y - from->y;
    float length = sqrtf(dx * dx + dy * dy);
    if (length < EPSILON)
        return; // 정지 상태, 무시

    float curr_vec_x = dx / length;
    float curr_vec_y = dy / length;

    // 평균 벡터 누적 (normalize는 나중에 필요 시)
    p->avg_vec_x += curr_vec_x;
    p->avg_vec_y += curr_vec_y;
    p->vec_count += 1;
}

coord route_get_coord_at(route p, int index) {
    if (!p || index < 0)
        return NULL;

    GList* node = g_list_nth(p->coords, index);
    if (!node)
        return NULL;

    return (coord)(node->data);
}

void route_update_average_vector_by_index(
    route p, int index_from, int index_to) {

    coord from = route_get_coord_at(p, index_from);
    coord to = route_get_coord_at(p, index_to);
    if (!from || !to) return;

    int dx = to->x - from->x;
    int dy = to->y - from->y;
    gfloat length = sqrtf(dx * dx + dy * dy);
    if (length == 0) return;

    p->avg_vec_x = (
        (p->avg_vec_x * p->vec_count) + (dx / length)) / (p->vec_count + 1);

    p->avg_vec_y = (
        (p->avg_vec_y * p->vec_count) + (dy / length)) / (p->vec_count + 1);

    p->vec_count += 1;
}

gboolean route_has_changed_by_index(
    route p, int index_from, int index_to, gfloat angle_threshold_deg) {

    gfloat angle = 0.0f;
    return route_has_changed_with_angle_by_index(
        p, index_from, index_to, angle_threshold_deg, &angle);
}

gboolean route_has_changed_with_angle_by_index(
    route p,
    int index_from,
    int index_to,
    gfloat angle_threshold_deg,
    gfloat* out_angle_deg) {

    coord from = route_get_coord_at(p, index_from);
    coord to = route_get_coord_at(p, index_to);
    if (!from || !to || !out_angle_deg) return FALSE;

    int dx = to->x - from->x;
    int dy = to->y - from->y;
    gfloat length = sqrtf(dx * dx + dy * dy);
    if (length == 0) {
        *out_angle_deg = 0.0f;
        return FALSE;
    }

    gfloat vec_x = dx / length;
    gfloat vec_y = dy / length;

    gfloat dot = (vec_x * p->avg_vec_x) + (vec_y * p->avg_vec_y);
    if (dot > 1.0f) dot = 1.0f;
    if (dot < -1.0f) dot = -1.0f;

    gfloat angle_rad = acosf(dot);
    *out_angle_deg = angle_rad * (180.0f / (gfloat)G_PI);
    return (*out_angle_deg > angle_threshold_deg);
}

gfloat route_calc_average_dir(route p, int history) {
    if (!p || history <= 0) return 0.0f;

    int len = route_length(p);
    if (len < 2) return 0.0f;

    int base_index = (len > history) ? (len - history - 1) : 0;
    int latest_index = len - 1;

    coord base = route_get_coord_at(p, base_index);
    coord latest = route_get_coord_at(p, latest_index);
    if (!base || !latest) return 0.0f;

    int dx = latest->x - base->x;
    int dy = latest->y - base->y;

    if (dx == 0 && dy == 0) return 0.0f; // 정지 상태

    float angle_rad = atan2f((float)dy, (float)dx);
    float angle_deg = angle_rad * (180.0f / (float)M_PI);

    return angle_deg;
}

route_dir_t route_calc_average_facing(route p, int history) {
    if (!p || history < 1) return ROUTE_DIR_UNKNOWN;

    int len = route_length(p);
    if (len < 2) return ROUTE_DIR_UNKNOWN;

    int last = len - 1;
    int from = last - history;
    if (from < 0) from = 0;

    coord c_from = route_get_coord_at(p, from);
    coord c_to   = route_get_coord_at(p, last);

    if (!c_from || !c_to) return ROUTE_DIR_UNKNOWN;

    int dx = c_to->x - c_from->x;
    int dy = c_to->y - c_from->y;

    // 방향 정규화
    if (dx > 1) dx = 1;
    if (dx < -1) dx = -1;
    if (dy > 1) dy = 1;
    if (dy < -1) dy = -1;

    for (int i = 1; i <= 8; ++i) {
        if (dx == ROUTE_DIRECTION_VECTORS[i][0] &&
            dy == ROUTE_DIRECTION_VECTORS[i][1]) {
            return (route_dir_t)i;
        }
    }

    return ROUTE_DIR_UNKNOWN;
}

route_dir_t calc_direction(const coord start, const coord goal) {
    if (!start || !goal) return ROUTE_DIR_UNKNOWN;

    int dx = goal->x - start->x;
    int dy = goal->y - start->y;

    if (dx > 1) dx = 1;
    if (dx < -1) dx = -1;
    if (dy > 1) dy = 1;
    if (dy < -1) dy = -1;

    for (int i = 1; i <= 8; ++i) {
        if (dx == ROUTE_DIRECTION_VECTORS[i][0] &&
            dy == ROUTE_DIRECTION_VECTORS[i][1]) {
            return (route_dir_t)i;
        }
    }

    return ROUTE_DIR_UNKNOWN;
}

coord direction_to_coord(route_dir_t dir) {
    if (dir < ROUTE_DIR_UNKNOWN || dir > ROUTE_DIR_DOWN_RIGHT)
        return NULL;

    return coord_new_full(ROUTE_DIRECTION_VECTORS[dir][0],
                          ROUTE_DIRECTION_VECTORS[dir][1]);
}

void route_insert(route p, int index, const coord c) {
    if (!p || index < 0 || index > g_list_length(p->coords)) return;
    coord cpy = coord_copy(c);
    p->coords = g_list_insert(p->coords, cpy, index);
}

void route_remove_at(route p, int index) {
    if (!p || index < 0 || index >= g_list_length(p->coords)) return;
    GList* nth = g_list_nth(p->coords, index);
    if (nth) {
        g_free(nth->data);
        p->coords = g_list_delete_link(p->coords, nth);
    }
}

void route_remove_value(route p, const coord c) {
    if (!p) return;
    for (GList* l = p->coords; l; l = l->next) {
        if (coord_equal(l->data, c)) {
            g_free(l->data);
            p->coords = g_list_delete_link(p->coords, l);
            break;
        }
    }
}

gboolean route_contains(const route p, const coord c) {
    if (!p) return FALSE;
    for (GList* l = p->coords; l; l = l->next)
        if (coord_equal(l->data, c)) return TRUE;
    return FALSE;
}

gint route_find(const route p, const coord c) {
    if (!p) return -1;
    int i = 0;
    for (GList* l = p->coords; l; l = l->next, ++i)
        if (coord_equal(l->data, c)) return i;
    return -1;
}

void route_slice(route p, int start, int end) {
    if (!p || start < 0 || end <= start) return;
    int len = g_list_length(p->coords);
    if (end > len) end = len;

    GList* new_list = NULL;
    for (int i = start; i < end; ++i) {
        coord c = g_list_nth_data(p->coords, i);
        if (c) new_list = g_list_append(new_list, coord_copy(c));
    }

    g_list_free_full(p->coords, g_free);
    p->coords = new_list;
}